gadjid/src/graph_operations/reachability.rs

```diff
@@ -0,0 +1,722 @@
+            WalkStatus::PD_BLOCK_AM => {
+                not_vas.insert(node);
+                poss_de.insert(node);
+            }
+            WalkStatus::PD_OPEN_AM => {
+                poss_de.insert(node);
+            }
+            _ => (),
+        }
+        let node_is_adjustment = z.contains(&node);
+
+        for (move_on_by, w, blocked) in get_next_steps(arrived_by, node, node_is_adjustment) {
+            let next = match walkstatus {
+                WalkStatus::Init => match move_on_by {
+                    Edge::Incoming => Some((move_on_by, w, WalkStatus::PD_OPEN_AM)),
+                    Edge::Outgoing => Some((move_on_by, w, WalkStatus::NON_CAUSAL)),
+                    Edge::Undirected => Some((move_on_by, w, WalkStatus::PD_OPEN_NAM)),
+                    _ => None,
+                },
+                WalkStatus::PD_OPEN_AM | WalkStatus::PD_BLOCK_AM => match move_on_by {
+                    Edge::Incoming | Edge::Undirected => match blocked {
+                        false => Some((move_on_by, w, walkstatus)),
+                        true => Some((move_on_by, w, WalkStatus::PD_BLOCK_AM)),
+                    },
+                    Edge::Outgoing if !blocked && matches!(walkstatus, WalkStatus::PD_OPEN_AM) => {
+                        Some((move_on_by, w, WalkStatus::NON_CAUSAL))
+                    }
+                    _ => None,
+                },
+                WalkStatus::PD_OPEN_NAM | WalkStatus::PD_BLOCK_NAM => match move_on_by {
+                    Edge::Incoming | Edge::Undirected => match blocked {
+                        false => Some((move_on_by, w, walkstatus)),
+                        true => Some((move_on_by, w, WalkStatus::PD_BLOCK_NAM)),
+                    },
+                    Edge::Outgoing if !blocked && matches!(walkstatus, WalkStatus::PD_OPEN_NAM) => {
+                        Some((move_on_by, w, WalkStatus::NON_CAUSAL))
+                    }
+                    _ => None,
+                },
+                WalkStatus::NON_CAUSAL if !blocked => Some((move_on_by, w, WalkStatus::NON_CAUSAL)),
+                _ => None,
+            };
+
+            if let Some(next) = next {
+                if !visited.contains(&next) {
+                    to_visit_stack.push(next);
+                }
+            }
+        }
+    }
+
+    assert!(not_amenable.is_subset(&not_vas));
+
+    (poss_de, not_amenable, not_vas)
+}
+
+/// Checks amenability of some DAG/CPDAG for a given set T of treatments and all response variables
+///
+/// Follows Algorithm 5 in https://doi.org/10.48550/arXiv.2402.08616
+///
+/// Returns tuple of:<br>
+/// - Set PD of possible descendants of T in G
+/// - Set NAM (Not AMenable) of nodes Y \notin T in G such that G is not amenable relative to (T, Y)
+pub fn get_pd_nam(truth_dag: &PDAG, t: &[usize]) -> (FxHashSet<usize>, FxHashSet<usize>) {
+    #[allow(non_camel_case_types)]
+    #[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+    enum Alg5WalkStatus {
```

Comment:
I think it makes it easier to understand the algo (as there are fewer possible transitions to consider) when reducing the walk stati to only the ones relevant to computing result sets we need.
The enum is only useful within that function, I think it would clutter more moving it outside.
I will consolidate naming.

Reply:
Agreed with keeping them local, changed naming so we have the PD = Possibly Directed / Possible Descendant overloaded mnemonic and no Alg5/6/7s :-)